File: codility/lessons/1 - iterations/binary gap/binary_gap.ts
// Source: https://app.codility.com/programmers/lessons/1-iterations/binary_gap/

function solution(N: number): number {
    const BINARY: string = convertNumberToBinaryString(N);
    const LONGEST_BINARY_GAP: number = findLongestBinaryGap(BINARY);
    return LONGEST_BINARY_GAP;
}

    function convertNumberToBinaryString(N: number): string {
        const BINARY: string = (N >>> 0).toString(2);
        return BINARY;
    }

    function findLongestBinaryGap(BINARY: string): number {
        let longestBinaryGap = 0;

        for (let index = {'i': 0, 'j': 0, 'k': 0}; Object.values(index).includes(-1) === false;) {
            const GAP_LENGTH: number = findGapLength(BINARY, index);
            longestBinaryGap = Math.max(GAP_LENGTH, longestBinaryGap);
        }

        return longestBinaryGap;
    }

        function findGapLength(BINARY: string, index: { i: number; j: number; k: number; } ): number {
            index.i = BINARY.indexOf('1', index.j);
            index.j = BINARY.indexOf('0', index.i + 1);
            index.k = BINARY.indexOf('1', index.j + 1);

            if (Object.values(index).includes(-1) === false) {
                const GAP_LENGTH: number = index.k - index.j;
                return GAP_LENGTH;
            }

            return 0;
        }

File: codility/lessons/1 - iterations/binary gap/binary_gap.cs
// Source: https://app.codility.com/programmers/lessons/1-iterations/binary_gap/

using System;
using System.Collections.Generic;


class Solution
{
    private string _binary;
    private int _longestBinaryGap;

    public int solution(int NUMBER)
    {
        _binary = _ConvertIntToBinaryString(NUMBER);
        _longestBinaryGap = _FindLongestBinaryGap();
        return _longestBinaryGap;
    }

        private string _ConvertIntToBinaryString(int NUMBER)
        {
            string binary = Convert.ToString(NUMBER, 2);
            return binary;
        }

        private int _FindLongestBinaryGap()
        {
            int longestBinaryGap = 0;

            for (List<int> index = new List<int>(new int[3]); !index.Contains(-1);)
            {
                int gapLength = _FindGapLength(index);
                longestBinaryGap = Math.Max(gapLength, longestBinaryGap);
            }

            return longestBinaryGap;
        }

            private int _FindGapLength(List<int> index)
            {
                index[0] = _binary.IndexOf('1', index[1]);
                if (index[0] != -1)
                {
                    index[1] = _binary.IndexOf('0', index[0] + 1);
                    if (index[1] != -1)
                    {
                        index[2] = _binary.IndexOf('1', index[1] + 1);
                        if (index[2] != -1)
                        {
                            int gapLength = index[2] - index[1];
                            return gapLength;
                        }
                        else
                            return 0;
                    }
                    else
                        return 0;
                }
                else
                    return 0;
            }
}
